## 1. Course Schedule

In [13]:
from bs4 import BeautifulSoup
import pandas as pd

# Load the file
with open("fall2023_courseSched.html", "r") as file:
    soup_schedule = BeautifulSoup(file, "html.parser")

In [14]:
# Find the div containing the course schedule
div = soup_schedule.find_all('li', {'class': 'row'})

course_data = []

for course in div: 
    span = course.find_all('span')
    number_section = span[0].text.replace('\u200b', '').strip('\n').split('\n')
    number_section = number_section[0]
    name = span[1].text.strip()
    instructor = span[2].text.strip()
    time = span[3].text.strip()
    # Append and update course schedule data list
    course_data.append([number_section, name, instructor, time])

# Convert data to pandas DataFrame 
schedule = pd.DataFrame(course_data, columns = ['Number-Section', 'Name', 'Instructor', 'Time'])

In [15]:
schedule.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 175 entries, 0 to 174
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Number-Section  175 non-null    object
 1   Name            175 non-null    object
 2   Instructor      175 non-null    object
 3   Time            175 non-null    object
dtypes: object(4)
memory usage: 5.6+ KB


In [16]:
# Splitting 'Number-Section' into 'Number' and 'Section' using regex
schedule[['Number', 'Section']] = schedule['Number-Section'].str.extract(r'([A-Za-z]+-[A-Za-z]+)\.([0-9]+)')

# Drop original 'Number-Section' column since data is stored in two seperate columns
schedule = schedule.drop('Number-Section', axis=1)

# Rearrange the columns
schedule = schedule[['Number', 'Section', 'Name', 'Instructor', 'Time']]
schedule.head()

Number Section                               Name      Instructor  \
0  CSCI-GA    1133                             PAC I*       CANCELLED   
1  CSCI-GA    1133                   PAC I Recitation       CANCELLED   
2  CSCI-GA    1170             Fundamental Algorithms  Yevgeniy Dodis   
3  CSCI-GA    1170  Fundamental Algorithms Recitation      Peter Hall   
4  CSCI-GA    1170             Fundamental Algorithms    Ernest Davis   

            Time  
0              -  
1              -  
2  T 4:55-6:55PM  
3  R 5:55-6:45PM  
4  T 4:55-6:55PM

In [17]:
# Replace 'TTh' with 'TR' for consistency
schedule['Time'] = schedule['Time'].str.replace('TTh ', 'TR', regex=False)
# Replace '-' with 'Not Scheduled' in 'Time' column
schedule['Time'] = schedule['Time'].replace('-', 'Not Scheduled')
# Ensure consistent format in 'Time' column
schedule['Time'] = schedule['Time'].str.replace('R ', 'Th ', regex=False)
# If there's any leftover 'TTh', replace them with 'TR'
schedule['Time'] = schedule['Time'].str.replace('TTh', 'Th ', regex=False)

# Replace 'STAFF' with 'Not Assigned' in 'Instructor' column
schedule['Instructor'] = schedule['Instructor'].replace('STAFF', 'Not Assigned')
# Drop rows where 'Instructor' is 'CANCELLED'
schedule = schedule[schedule['Instructor'] != 'CANCELLED']

# Replace 'NaN' values with 'Unknown'
schedule[['Number', 'Section']] = schedule[['Number', 'Section']].fillna('Unknown')


In [18]:
schedule.head()

Number Section                               Name      Instructor  \
2  CSCI-GA    1170             Fundamental Algorithms  Yevgeniy Dodis   
3  CSCI-GA    1170  Fundamental Algorithms Recitation      Peter Hall   
4  CSCI-GA    1170             Fundamental Algorithms    Ernest Davis   
5  CSCI-GA    1170  Fundamental Algorithms Recitation      Eli Goldin   
6  CSCI-GA    1170  Fundamental Algorithms Recitation      Sihyun Lee   

             Time  
2   T 4:55-6:55PM  
3  Th 5:55-6:45PM  
4   T 4:55-6:55PM  
5   F 4:55-5:45PM  
6  Th 5:55-6:45PM

In [19]:
schedule.tail()

Number  Section                                               Name  \
170  CSCI-UA     0480  Special Topics: Agile Software Development and...   
171  CSCI-UA     0480  Special Topics:\n                             ...   
172  Unknown  Unknown                             Undergraduate Research   
173  Unknown  Unknown                                         Internship   
174  Unknown  Unknown                                  Independent Study   

         Instructor              Time  
170  Amos Bloomberg  Th  12:30-1:45PM  
171    Lerrel Pinto   Th  3:30-4:45PM  
172    Not Assigned     Not Scheduled  
173    Not Assigned     Not Scheduled  
174    Not Assigned     Not Scheduled

In [20]:
schedule.sample(5)

Number Section                                               Name  \
35   CSCI-GA    2820                     DevOps and Agile Methodologies   
110  CSCI-UA    0101                          Intro To Computer Science   
45   CSCI-GA    3033  Special Topics:\n                             ...   
128  CSCI-UA    0102                       Data Structures - Recitation   
5    CSCI-GA    1170                  Fundamental Algorithms Recitation   

          Instructor            Time  
35      John Rofrano   T 4:55-6:55PM  
110        Sana Odeh  MW 4:55-6:10PM  
45    Joseph Bonneau   T 7:10-9:10PM  
128  Anirudh Nistala   F 2:00-3:15PM  
5         Eli Goldin   F 4:55-5:45PM

## 2. Course Catalog

In [22]:
# Load the file
with open("course_catalog.html", "r") as file:
    soup_catalog = BeautifulSoup(file, "html.parser")

In [23]:
# Find the div containing the course schedule
div = soup_catalog.find_all('li', {'class': 'col-sm-12'})

course_catalog = []

for catalog in div: 
    span = catalog.find_all('p',{'class': 'bold'})
    number=''.join(span[0].text.replace('\n','').strip().split(' ')[0])
    prereqs = span[2].text.split('Prerequisites:')[1]
    points = span[1].text.replace('\n','').strip().split(' ')[0]
    # Append and update course schedule data list
    course_catalog.append([number, prereqs, points])

# Convert data to pandas DataFrame 
catalog = pd.DataFrame(course_catalog, columns=['Number', 'Prereqs', 'Points'])

In [24]:
catalog.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96 entries, 0 to 95
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   Number   96 non-null     object
 1   Prereqs  96 non-null     object
 2   Points   96 non-null     object
dtypes: object(3)
memory usage: 2.4+ KB


In [25]:
# Splitting 'Number' into 'Number' and 'Section' using regex
# Splitting the column is necessary for when we merge both dataframes together
catalog[['Number', 'Section']] = catalog['Number'].str.split('.', expand=True)

# Rearrange the columns
catalog = catalog[['Number', 'Prereqs', 'Points']]

In [26]:
catalog.head()

Number                                            Prereqs Points
0  CSCI-GA                                               None      4
1  CSCI-GA           CSCI-GA 1133 or departmental permission.      4
2  CSCI-GA   At least one year of experience with a high-l...      3
3  CSCI-GA                                               None      3
4  CSCI-GA   Students taking this class should already hav...      3

In [27]:
catalog.tail()

Number                                            Prereqs Points
91   CSCI-UA   Restricted to declared computer science major...      1
92   CSCI-UA   Restricted to declared computer science major...      1
93   CSCI-UA   Permission of the department. Does not satisf...      1
94   CSCI-UA   Permission of the department. Does not satisf...      1
95  FRSEM-UA   Some programming experience in Python, Java, ...      4

In [28]:
catalog.sample(5)

Number                                            Prereqs Points
53  CSCI-GA   Permission of the instructor for master’s stu...      4
82  CSCI-UA        Computer Systems Organization (CSCI-UA 201)      4
63  CSCI-UA   Three years of high school mathematics or equ...      4
35  CSCI-GA   1. Undergraduate Operating Systems or Compute...      3
45  CSCI-GA             Prerequisites vary according to topic.      3

## 3. Merge both dataframes

In [29]:
df = pd.merge(schedule, catalog, on='Number', how='left')

# reorder the columns as specified
df = df[['Number', 'Name', 'Instructor', 'Time', 'Prereqs', 'Points']]
df

Number                                               Name  \
0     CSCI-GA                             Fundamental Algorithms   
1     CSCI-GA                             Fundamental Algorithms   
2     CSCI-GA                             Fundamental Algorithms   
3     CSCI-GA                             Fundamental Algorithms   
4     CSCI-GA                             Fundamental Algorithms   
...       ...                                                ...   
6922  CSCI-UA  Special Topics:\n                             ...   
6923  CSCI-UA  Special Topics:\n                             ...   
6924  Unknown                             Undergraduate Research   
6925  Unknown                                         Internship   
6926  Unknown                                  Independent Study   

          Instructor             Time  \
0     Yevgeniy Dodis    T 4:55-6:55PM   
1     Yevgeniy Dodis    T 4:55-6:55PM   
2     Yevgeniy Dodis    T 4:55-6:55PM   
3     Yevgeniy Dodis    T 4:55-6:55PM   
4     Yevgeniy Dodis    T 4:55-6:55PM   
...              ...              ...   
6922    Lerrel Pinto  Th  3:30-4:45PM   
6923    Lerrel Pinto  Th  3:30-4:45PM   
6924    Not Assigned    Not Scheduled   
6925    Not Assigned    Not Scheduled   
6926    Not Assigned    Not Scheduled   

                                                Prereqs Points  
0                                                  None      4  
1              CSCI-GA 1133 or departmental permission.      4  
2      At least one year of experience with a high-l...      3  
3                                                  None      3  
4      Students taking this class should already hav...      3  
...                                                 ...    ...  
6922   Permission of the department. Does not satisf...      1  
6923   Permission of the department. Does not satisf...      1  
6924                                                NaN    NaN  
6925                                                NaN    NaN  
6926                                                NaN    NaN  

[6927 rows x 6 columns]

In [30]:
pd.set_option('display.max_rows', 200) # Shows the entire dataframe

In [31]:
df = df.groupby(['Number', 'Name', 'Instructor', 'Time']).agg({
    'Prereqs': lambda x: list(x.unique()),
    'Points': 'first'
}).reset_index()
df

Number                                               Name  \
0    CSCI-GA                          Advanced Database Systems   
1    CSCI-GA  Advanced Topics in Numerical Analysis: Computa...   
2    CSCI-GA  Advanced Topics in Numerical Analysis: Monte C...   
3    CSCI-GA            Applied Cryptography & Network Security   
4    CSCI-GA                            Artificial Intelligence   
5    CSCI-GA                   Big Data Application Development   
6    CSCI-GA                                  Computer Graphics   
7    CSCI-GA                                    Computer Vision   
8    CSCI-GA                   Data Communications and Networks   
9    CSCI-GA                                   Database Systems   
10   CSCI-GA                                      Deep Learning   
11   CSCI-GA                                  Deep Learning Lab   
12   CSCI-GA                         Design and Innovation Lab*   
13   CSCI-GA                             Design and Innovation*   
14   CSCI-GA                     DevOps and Agile Methodologies   
15   CSCI-GA                     DevOps and Agile Methodologies   
16   CSCI-GA                 DevOps and Agile Methodologies Lab   
17   CSCI-GA                 DevOps and Agile Methodologies Lab   
18   CSCI-GA                   Foundations Of Machine Learning*   
19   CSCI-GA    Foundations of Networks and Mobile Systems Lab*   
20   CSCI-GA    Foundations of Networks and Mobile Systems Lab*   
21   CSCI-GA        Foundations of Networks and Mobile Systems*   
22   CSCI-GA        Foundations of Networks and Mobile Systems*   
23   CSCI-GA                             Fundamental Algorithms   
24   CSCI-GA                             Fundamental Algorithms   
25   CSCI-GA                  Fundamental Algorithms Recitation   
26   CSCI-GA                  Fundamental Algorithms Recitation   
27   CSCI-GA                  Fundamental Algorithms Recitation   
28   CSCI-GA                          Heuristic Problem Solving   
29   CSCI-GA                      Honors Analysis of Algorithms   
30   CSCI-GA                       Honors Programming Languages   
31   CSCI-GA                  Information Technology Projects *   
32   CSCI-GA                       Introduction To Cryptography   
33   CSCI-GA                                   Machine Learning   
34   CSCI-GA        Mathematical Techniques For CS Applications   
35   CSCI-GA                        Natural Language Processing   
36   CSCI-GA                                Numerical Methods I   
37   CSCI-GA                                  Operating Systems   
38   CSCI-GA                        PhD Seminar: Cryptography**   
39   CSCI-GA                      PhD Seminar: Formal Methods**   
40   CSCI-GA                              Programming Languages   
41   CSCI-GA                              Programming Languages   
42   CSCI-GA                   Programming Languages Recitation   
43   CSCI-GA                   Programming Languages Recitation   
44   CSCI-GA                   Programming Languages Recitation   
45   CSCI-GA                    Realtime and Big Data Analytics   
46   CSCI-GA                               Scientific Computing   
47   CSCI-GA    Special Topics in CS: Bayesian Machine Learning   
48   CSCI-GA  Special Topics in CS: Bayesian Machine Learnin...   
49   CSCI-GA  Special Topics:\n                             ...   
50   CSCI-GA  Special Topics:\n                             ...   
51   CSCI-GA  Special Topics:\n                             ...   
52   CSCI-GA  Special Topics:\n                             ...   
53   CSCI-GA  Special Topics:\n                             ...   
54   CSCI-GA  Special Topics:\n                             ...   
55   CSCI-GA  Special Topics:\n                             ...   
56   CSCI-GA  Special Topics:\n                             ...   
57   CSCI-GA  Special Topics:\n                             ...   
58   CSCI-GA  Special Topics:\n                      

## 4. Conclusion: observations, ouliers and discrepancies 

I noticed that merging the dataframe has resulted in a very large dataset (6927 rows and 6 columns). By looking at the first 5 rows, I have noticed the columns are identical other than the 'Prereqs' and 'Points' column. In a scenario like this one, I decided to group the columns by number, name, instructor while assigning the prereqs column to only include the unique elements of the repeated sections as well as assigning the first element of points for the points column. 

Although this may not be the most accurate way to do so, we can see that the shape of the dataframe reduced immensely and provoides a more concise overview of the course/catalog schedule without necessarily 'dropping' or emitting any courses. 

Additionally, the merging of both dataframes has caused a few columns to be marked as NaN values, as seen between rows 158-164. This is most likely due to the fact that both dataframes contained common blank spaces (such as Number, Instructor or Time). It is also worth noting that these courses are considered to be more 'niche' and may not be applicable to the geenral population of the NYU community. Hence, we can either decided to drop/remove these courses from the schedule and possibly store them in a seperate table where specific information/details of the course could be inputted manually.

Left join: all the elements from the second table that match the key values contained in first table are merged. Elements from the second table that are not matched by key values in the first table are not part of the resulting merged dataframe.